flywheel-io/core
# Rerun rule-spawned jobs on replaced files

Closes #1081 
Closes #878

_Copied from `test_rules_rerun_after_file_replace` function sig:_

Always run jobs from rules where at least one of the inputs were "replaced" during the upload.

When a file is uploaded to a container that already has a file with the same name, the original file is completely replaced with a new one in mongo. 
      * Except when the upload type is "reaper", and the hash is the same. In that case the file upload is ignored.
      ** See comment in code for discussion about created/replaced timestamp.

In the rule evaluation step, the rules are evaluated against the container's state before the upload (set A) and after (set B). Jobs in set B that are not in set A are queued if no files were replaced during the upload.

If files were replaced, any job in set A that has a replaced file as input is also queued.

### Breaking Changes
None, although rules will spawn jobs when they had not before

### Affected Components
File Upload, Reaper Upload, Rules

### Test Plan Overview 
  - Upload files via the different placer endpoints, ensure entire file object is replaced, created date stays the same, replaced date is added
  - Upload files via `/upload/reaper`. Make sure replaces of the same file with the same hash are ignored.
  - Ensure expected rules trigger as if the file was uploaded for the first time
  - Test reaper uploads on ObjS
  
### Checklist
- [x] Integration/Unit tests
- [x] Swagger updates and changes (just for replaced field if it sticks around)
- [x] Test Plan Documentation

### CCs
@ryansanford @gsfr @tcbtcb
